README.md
# k8s

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-07-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-07-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/k8s/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/k8s/compare/0.1.0...0.1.1


test/k8s/info_test.clj
(ns k8s.info-test
  (:require [k8s.info :as sut]
            [matcho.core :as matcho]
            [clojure.test :refer :all]))

(deftest secrets-test
  (matcho/match
   (sut/resolve-secrets
    {} {:spec   {:prop {:valueFrom {:secretKeyRef {:name "Secrets" :key "Key1"}}}}
        :nested {:attr1 {:natr2 {:valueFrom {:secretKeyRef {:name "DB" :key "password"}}}
                         :natr3 {:valueFrom {:configMapKeyRef {:name "CM" :key "cfg"}}}}}
        :vector [{:prop {:valueFrom {:secretKeyRef {:name "Secrets" :key "Key1"}}}}
                 {:prop {:valueFrom {:secretKeyRef {:name "Secrets" :key "Key2"}}}}]})
   {:secretKeyRef {:Secrets {:Key1 [[:spec :prop] [:vector 0 :prop]]
                        :Key2 [[:vector 1 :prop]]}
                   :DB {:password [[:nested :attr1 :natr2]]}}
    :configMapKeyRef {:CM {:cfg [[:nested :attr1 :natr3]]}}}))

(def ctx {:base-url "http://locahost:8001"})

(deftest url-builder-test 

  (is (= (sut/build-url ctx {:kind "PersistentVolumeClaim" :apiVersion "v1"})
         "http://locahost:8001/api/v1/persistentvolumeclaims"))

  (is (= (sut/build-url ctx {:kind "PersistentVolumeClaim" :apiVersion "apiextensions.k8s.io/v1beta1"} {:labelSelector "system in (c3)"})
         "http://locahost:8001/apis/apiextensions.k8s.io/v1beta1/persistentvolumeclaims?labelSelector=system+in+%28c3%29"))

  (is (= (sut/build-url ctx {:kind "PersistentVolumeClaim" :apiVersion "v1" :ns "test"})
       "http://locahost:8001/api/v1/namespaces/test/persistentvolumeclaims"))

  (is (= (sut/build-url ctx {:kind "PersistentVolumeClaim" :apiVersion "v1" :ns "test"})
       "http://locahost:8001/api/v1/namespaces/test/persistentvolumeclaims"))

  )


test/k8s/jetty_test.clj
(ns k8s.jetty-test
  (:require [k8s.jetty :as sut]
            [clojure.test :refer :all]))


(deftest jetty-test


  (def cl (sut/client {:insecure? true}))




  )



project.clj
(defproject k8s "0.1.0-SNAPSHOT"
  :description "k8s client for clojure"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.9.0-alpha16"]
                 [cheshire "5.7.1"]
                 [clj-json-patch "0.1.4"]
                 [org.eclipse.jetty.http2/http2-client "9.4.7.v20170914"]
                 [org.eclipse.jetty.websocket/websocket-client "9.4.7.v20170914"]
                 [hiccup "1.0.5"]
                 [ch.qos.logback/logback-classic "1.2.2"]
                 [http-kit "2.2.0"]
                 [clj-yaml "0.4.0"]
                 [clj-jwt "0.1.1"]
                 [clj-time "0.13.0"]
                 [inflections "0.13.0"]
                 [org.mortbay.jetty.alpn/alpn-boot "8.1.9.v20160720"]
                 [matcho "0.1.0-RC5"]])


src/k8s/jetty.clj
(ns k8s.jetty
  (:import [java.net URLEncoder]
           [java.util Base64]
           [java.nio ByteBuffer]
           [org.eclipse.jetty.client HttpClient]
           [org.eclipse.jetty.websocket.api WebSocketListener Session WriteCallback RemoteEndpoint]
           [org.eclipse.jetty.websocket.client WebSocketClient ClientUpgradeRequest]
           [org.eclipse.jetty.util.ssl SslContextFactory]
           [org.eclipse.jetty.http2.api Stream]
           [org.eclipse.jetty.http2.api.server ServerSessionListener]
           [org.eclipse.jetty.http2.frames DataFrame HeadersFrame]

           [org.eclipse.jetty.http MetaData HttpFields HttpURI HttpVersion]
           [org.eclipse.jetty.util FuturePromise Callback]
           [java.util.concurrent TimeUnit]
           [org.eclipse.jetty.http2.client HTTP2Client]

           [java.nio.charset Charset])
  (:require [clojure.string :as str]
            [cheshire.core :as json]))

(defn ssl-context []
  (let [sec (SslContextFactory.)]
    (.setTrustAll sec true)
    (.setValidateCerts sec false)
    (.start sec)
    sec))

(defonce ssl-ctx (atom nil))

(defn ensure-ssl-context []
  (if-let [ctx @ssl-ctx]
    ctx
    (reset! ssl-ctx (ssl-context))))


(defn client [{insecure? :insecure?}]
  ;; httpClient.setFollowRedirects(false);
  (let [cl (if insecure?
             (HttpClient. (ensure-ssl-context))
             (HttpClient.))]
    (.start cl)
    cl))

(defn parse-body [fmt body]
  (cond
    (and (= fmt "application/json") body)
    (json/parse-string body keyword)
    :else body))

(defn request
  [^HttpClient client
   {uri :uri
    params :params
    headers :headers
    method :method :as opts}]

  (let [req (.newRequest client uri)]
    (.method req (or "GET" method))

    (when params
      (doseq [[k v] params]
        (.param req (name k) v)))

    (when headers
      (doseq [[k v] headers]
        (.header req (name k) v)))

    (let [resp (.send req)
          headers (reduce (fn [acc field]
                            (assoc acc (keyword (.getName field)) (.getValue field)))
                          {} (.getHeaders resp))
          body (.getContentAsString resp)]
      {:status (.getStatus resp)
       :headers headers
       :body (parse-body (:Content-Type headers) body)})))

(defn ws-client [{insecure? :insecure?}]
  (let [cl (if insecure?
             (WebSocketClient. (ensure-ssl-context))
             (WebSocketClient.))]
    (.start cl)
    cl))

(def utf (Charset/forName "utf-8"))

(defn ws-listener
  []
  (let [p (promise)
        sb (StringBuilder.)]
    [p (reify WebSocketListener
       (onWebSocketConnect [this session]
         (println "Session" session))

       (onWebSocketText [this message]
         (println "on text not impl" message))

       (onWebSocketBinary [this bytes offset len]
         ;; (println "msg:" :len len :off offset (String. bytes))
         (.append sb (String. bytes)))

       (onWebSocketError [this cause]
         (deliver p {:error cause
                     :body (.toString sb)}))

       (onWebSocketClose [this status-code reason]
         (deliver p {:body (.toString sb)
                     :status status-code
                     :reason reason})))]))

(defn ws-request
  [^WebSocketClient client
   {uri :uri
    params :params
    headers :headers
    method :method :as opts}]

  (let [req (ClientUpgradeRequest.)]
    (.setMethod req (or "GET" method))

    (when params
      (doseq [[k v] params]
        (.setParam req (name k) v)))

    (when headers
      (doseq [[k v] headers]
        (.setHeader req (name k) v)))

    (let [[p listener]  (ws-listener)]
      (.connect client listener (java.net.URI. uri) req)
      @p)))


(defn http2-client
  [{insecure? :insecure?
    host :host
    port :port}]
  (let [cl (HTTP2Client.)]
    (.start cl)
    cl))

(defn http2-listener []
  (let [p (promise)
        sb (StringBuilder.)]
    [p (proxy [org.eclipse.jetty.http2.api.Stream$Listener$Adapter] []
         (onData [^Stream stream
                  ^DataFrame frame
                  ^Callback callback]
           (let [bs (byte-array (-> frame (.getData) (.remaining)))];)])
             (-> frame (.getData) (.get bs))
             (println "Data:" (String. bs))
             (.succeeded callback))))]))

(defn http2-request [client {uri :uri
                             host :host
                             port :port
                             method :method}]
  (let [method  (or "GET" method)
        huri    (HttpURI. uri)
        fields  (HttpFields.)
        req     (org.eclipse.jetty.http.MetaData$Request.
                 ^String method
                 ^HttpURI huri
                 HttpVersion/HTTP_2
                 ^HttpFields fields)
        headers (HeadersFrame. req nil true)
        [p listener] (http2-listener)
        pr (FuturePromise.)]
    (.connect client
              ;; ^SSLContextFactory (ensure-ssl-context)
              ^java.net.InetSocketAddress (java.net.InetSocketAddress. host (or port 80))
              (org.eclipse.jetty.http2.api.server.ServerSessionListener$Adapter.)
              ^FuturePromise pr)
    (let [session (.get pr 5 TimeUnit/SECONDS)]
      (.newStream session headers (FuturePromise.) listener)
      session)))

(comment

  (.start http)

  (def result
    (-> (.newRequest http "https://<api-server>/api/v1/namespaces/default/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz")
        (.method "GET")
        (.header "Authorization"
                 "Bearer <token>")
        (.send)))

  (.getStatus result)
  (.toString result)
  (.getBody result)

  (def cl (client {:insecure? true}))

  (def token "????")
  (def api-server "????")

  (def rs (request cl
             {:uri "https://<api-server>/api/v1/namespaces/default/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz"
              :method "GET"
              :headers {"Authorization" (str "Bearer " token)}}))
  (:body rs)

  (request cl
           {:uri "https://<api-server>/api/v1/namespaces/xxx/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz"
            :method "GET"
            :headers {"Authorization" (str "Bearer " token)}})

  (:headers rs)
  (:body rs)

  (def wws (ws-client {:insecure? true}))
  (def wres
    (ws-request wws
                {:uri (str "wss://<api-server>/api/v1/namespaces/default/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz/exec?"
                           "command=ls&command-lah"
                           "&container=pg&stderr=true&stdout=true")
                 :method "POST"
                 :headers {"Authorization" (str "Bearer " token)}}))

  wres



  (def http2 (http2-client {:insecure? true :host "localhost" :port 8001}))

  http2

  (def s
    (http2-request http2 {:uri "http://localhost:8001/api/v1/pods?watch=true"
                          :host "localhost"
                          :port 8001
                          :method "GET"}))

  s



  )


src/k8s/watch.clj
(ns k8s.watch)
;;   (ns k8s.watch
;;   (:require [http.async.client :as http]
;;             [http.async.client.cert :as cert]
;;             [clojure.tools.logging :as log]
;;             [clojure.string :as str]
;;             [cheshire.core :as json])
;;   (:import
;;    com.fasterxml.jackson.core.JsonParseException
;;    (javax.net.ssl
;;     KeyManagerFactory
;;     SSLContext
;;     HttpsURLConnection
;;     TrustManager
;;     X509TrustManager)))

;; (defonce state (atom {}))

;; (defrecord BlindTrustManager
;;     []
;;   X509TrustManager
;;   (checkClientTrusted [this chain auth-type] nil)
;;   (checkServerTrusted [this chain auth-type] nil)
;;   (getAcceptedIssuers [this] nil))

;; (defn #^SSLContext ssl-context
;;   []
;;   (let [ctx (SSLContext/getInstance "TLS")
;;         trust-managers (into-array javax.net.ssl.X509TrustManager
;;                                    (list (new BlindTrustManager)))]
;;     (.init ctx nil trust-managers nil)
;;     ctx))



;; (defn update-version [rt v]
;;   (when v
;;     (let [v (cond (string? v) (Integer/parseInt v)
;;                   (number? v) v)]
;;       (swap! state update-in [:requests rt :version]
;;              (fn [x] (if x (max x v) v))))))

;; (defn reset-version [rt]
;;   (swap! state assoc-in [:requests rt :version] nil))

;; (defn get-version [rt]
;;   (get-in @state [:requests rt :version]))


;; (defn set-request-data [rt k v]
;;   (assoc-in @state [:requests rt k] v))

;; (defn get-request-data [rt k]
;;   (get-in @state [:requests rt k]))

;; (defn *parse-json-stream [acc s]
;;   (if (str/includes? s "\n")
;;     (let [[json-str rest-msg] (str/split s #"\n" 2)]
;;       (try (*parse-json-stream (conj acc (json/parse-string json-str keyword)) rest-msg)
;;            (catch com.fasterxml.jackson.core.JsonParseException e
;;              (println "JSON ERROR WHILE PARSE" s)
;;              [s acc])))
;;     [s acc]))

;; (*parse-json-stream [] "{\"q\":2}\n{\"q\":1}\n")

;; (defn parse-json-stream [rt st]
;;   (let [[rest-of-msg resources]
;;         (*parse-json-stream  [] (str (or (get-request-data rt :message) "") (-> st .toByteArray String.)))]

;;     (set-request-data rt :message rest-of-msg)
;;     resources))

;; (defn mk-on-change [opts retry]
;;   (fn [st body]
;;     (doseq [res (parse-json-stream (:resource opts) body)]
;;       (update-version (:resource opts) (get-in res [:object :metadata :resourceVersion]))
;;       ;; sometimes version is compacted :(
;;       (if (and (= "ERROR" (:type res)) (= "Expired" (get-in res [:object :reason])))
;;         (do (println "OUTDATE version:" res)
;;             (reset-version (:resource opts))
;;             (retry))
;;         (do (println opts " for v:" (get-in res [:object :metadata :resourceVersion]))
;;             (println "->" (:handler opts)))))
;;     [body :continue]))

;; (defn build-watch-query
;;   [{kube-url :kube-url}
;;    {res :resource n :ns api :apiVersion}]
;;   {:url (str kube-url "/apis/" api "/namespaces/" n "/" (name res)) 
;;    :query  {:watch true}})

;; (defn do-watch [env client opts]
;;   (let [v (get-version (:resource opts)) 
;;         q (build-watch-query env opts)
;;         q (if v (assoc-in q [:query :resourceVersion] v) q)
;;         on-change (mk-on-change opts #(do-watch env client opts))]
;;     (println "Watch " opts " from " v)
;;     (->> (http/request-stream client :get (:url q) on-change
;;                               :headers {"Authorization" (str "Bearer " (:kube-token env))}
;;                               :insecure? true
;;                               :timeout 300000
;;                               :query (:query q))
;;          (assoc opts :request)
;;          (swap! state assoc-in [:requests (:resource opts)]))))

;; (defn start [{env :env cfg :watch}]
;;   (println "Start watching resources")
;;   (if-let [c (:client @state)]
;;     (.close c)
;;     (swap! state dissoc :client :requests))

;;   (let [client (http/create-client :ssl-context (ssl-context))]
;;     (swap! state assoc :client client)
;;     (doseq [res-cfg (:resources cfg)]
;;       (#'do-watch env client res-cfg))))

;; (defn supervisor [env]
;;   (future
;;     (println "Start supervisor")
;;     (while (not (:stop @state))
;;       (let [client (:client @state)]
;;         (if (.isClosed client)
;;           (do (println "Client closed, run it")
;;               (start))
;;           (doseq [[rt {req :request :as opts}] (:requests @state)]
;;             (cond (realized? (:error req))
;;                   (do (println "ERRORED:" opts)
;;                       (#'do-watch env client (dissoc opts :request)))
;;                   (realized? (:done req))
;;                   (do (println "DONE:" opts)
;;                       (#'do-watch env client (dissoc opts :request))))))
;;         (Thread/sleep 10000)))
;;     (println "Stopping supervisor")
;;     (swap! state :stop false)))


;; (comment
;;   (:version (:repositories (:requests @state)))
;;   (swap! state :stop true)
;;   (swap! state :stop false)
;;   state

;;   (supervisor {:kube-url "http://localhost:8001"})
;;   (start
;;    {:env {:kube-url "http://localhost:8001"}
;;     :watch {:timeout 5000
;;             :resources [{:handler :ci3.watch/repository
;;                          :apiVersion "ci3.io/v1"
;;                          :resource :repositories
;;                          :ns "default"}
;;                         {:handler :ci3.watch/build
;;                          :apiVersion "ci3.io/v1"
;;                          :resource :builds
;;                          :ns "default"}]}}))



src/k8s/core.clj
(ns k8s.core
  (:import [java.util Base64]
           (java.nio ByteBuffer)
           (org.eclipse.jetty.websocket.api WebSocketListener
                                            Session
                                            WriteCallback RemoteEndpoint)
           [org.httpkit.client SslContextFactory MultipartEntity]
           [org.eclipse.jetty.websocket.client
            WebSocketClient
            ClientUpgradeRequest])
  (:require
   [org.httpkit.client :as http-client]
   [clj-json-patch.core :as patch]
   [clojure.walk :as walk]
   [cheshire.core :as json]
   [inflections.core :as inflections]
   [clojure.string :as str]))

(def default-headers
  (if-let [token (System/getenv "KUBE_TOKEN")]
    {"Authorization" (str "Bearer " token)}
    {}))

(def kube-url (or (System/getenv "KUBE_URL") "http://localhost:8001"))

(defn url [pth]
  (str kube-url "/" pth))

(defn curl [pth]
  (let [res @(http-client/get
              (url pth)
              {:headers default-headers :insecure? true})]
    (-> res :body)))

(defn base64-decode [s]
  (if s
    (String. (.decode (Base64/getDecoder) s))
    nil))

(defn base64-encode [s]
  (if s
    (String. (.encode (Base64/getEncoder) (.getBytes s)))
    nil))

(defn secret [name key]
  (let [cfg {:apiVersion "v1" :ns "default"}]
    (->
     @(http-client/get
       (str kube-url  "/api/v1/namespaces/default/secrets/" name)
       {:insecure? true
        :headers (merge default-headers {"Content-Type" "application/json-patch+json"})})
     :body (json/parse-string keyword)
     :data key
     base64-decode)))

(defn resolve-secrets [res]
  (walk/postwalk
   (fn [x]
     (if (and (map? x) (contains? x :valueFrom))
       (if-let [{nm :name k :key} (get-in x [:valueFrom :secretKeyRef])]
         (if-let [res (secret nm (keyword k))]
           res
           (do (println "Could not resolve secret " k " from " nm) x))
         x)
       x)) res))

(defn to-query-string [m]
  (->> m
       (mapv (fn [[k v]]
               (str (name k) "=" (java.net.URLEncoder/encode (str v)))))
       (str/join "&")))

;; (to-query-string {:a 1 :b "x in (z)"})

(defn pluralize [x]
  (inflections/plural x))

(defn resource-url [cfg & parts]
  (println "CFG" cfg)

  (let [[path params] (if (map? (last parts))
                        [(butlast parts) (last parts)]
                        [parts nil])]
    (url (str (if (re-matches  #"^v[.0-9]+$" (:apiVersion cfg))
                    "api" "apis")
                  "/" (:apiVersion cfg)
                  (when-let [ns (or (:ns cfg) (get-in cfg [:metadata :namespace]))]
                    (str "/namespaces/" ns))
                  (str "/" (pluralize (str/lower-case (:kind cfg))))
                  (when (seq? path)
                    (str "/" (str/join "/" path)))
                  (when params (str "?" (to-query-string params)))))))

(resource-url {:kind "PersistentVolumeClaim" :apiVersion "v1"})
(resource-url {:kind "PersistentVolumeClaim" :apiVersion "apiextensions.k8s.io/v1beta1"} {:labelSelector "system in (c3)"})

(resource-url {:kind "PersistentVolumeClaim" :apiVersion "v1" :ns "test"})
(resource-url {:kind "PersistentVolumeClaim" :apiVersion "v1" :ns "test"})


(defn query [cfg & pth]
  (println  (apply resource-url cfg pth))
  (let [res @(http-client/get
              (apply resource-url cfg pth)
              {:headers (merge default-headers {"Content-Type" "application/json"})
               :insecure? true})]
    (-> res
     :body
     (json/parse-string keyword)
     (resolve-secrets))))

(defn do-find [cfg]
  (let [res (query cfg (or (:id cfg) (get-in cfg [:metadata :name])))]
    (when-not (= "Failure" (:status res))
      res)))

(defn create [res]
  (let [uri (resource-url res)]
    (println "POST" res)
    (-> @(http-client/post
          uri
          {:body (json/generate-string (walk/stringify-keys res))
           :insecure? true
           :headers (merge default-headers {"Content-Type" "application/json"})})
        :body
        (json/parse-string))))

(defn delete-collection [opts]
  (let [uri   (resource-url opts (select-keys opts [:labelSelector ]))]
    (println "DELETE " uri)
    (-> @(http-client/delete
          uri
          {:headers (merge default-headers {"Content-Type" "application/json"})
           :insecure? true})
        :body
        (json/parse-string))))

;; ReplicationController, ReplicaSet, StatefulSet, DaemonSet, and Deployment

(defn delete [res & [delete-opts]]
  (-> @(http-client/delete
        (str (resource-url res (get-in res [:metadata :name])))
        (cond-> {:headers (merge default-headers {"Content-Type" "application/json"})
                 :insecure? true}
          delete-opts (assoc :body (json/generate-string (walk/stringify-keys delete-opts)))))
      :body
      (json/parse-string)))

(defn get-json-patch [res nres]
  (println res)
  (let [-res (walk/stringify-keys (-> res
                                      (assoc :metadata (dissoc (:metadata res) :uid :resourceVersion :creationTimestamp :selfLink))
                                      (dissoc :status :selfLink)))
        -nres (walk/stringify-keys nres)]
    (patch/diff -res -nres)))

(defn patch [nres]
  (println "PATCH" (do-find nres))
  (if-let [res (do-find nres)]
    (let [diff  (get-json-patch res nres)
          uri (resource-url nres (or (:id nres) (get-in nres [:metadata :name])))]
      (println "PATCH" diff)
      (->
       @(http-client/patch
         uri
         {:body (json/generate-string diff)
          :insecure? true
          :headers (merge default-headers {"Content-Type" "application/json-patch+json"})})
       :body
       (json/parse-string keyword)))
    (create nres)))

(comment
  ;; (map :resourceVersion (map :metadata (:items (list cfg :repositories))))
  ;; (map :metadata)

  ;; (do-find cfg :repositories "ci3")

  ;; (resolve-secrets {:tar{:foo {:valueFrom {:secretKeyRef {:name "ci3" :key "bbKey"}}}}})
  ;; (resolve-secrets (do-find cfg :repositories "ci3"))

  ;; (patch cfg :repositories "ci3-chart" {:webhook nil})

  ;; (secret "ci3" :mySecret)
  ;; (secret "secrets" :github_token)

  ;; (secret "bitbucket" :oauthConsumerSecret)

  ;; (do-find cfg :builds "ci3-build-6")

  ;; (patch cfg :builds "test-1" {:status "changed"})

  ;; (delete cfg :repositories "ci3")


  ;; (curl {} "api/v1/namespaces/default/pods/aitem-hook-test-d40a2375646990b2dec75e80cf97ce5a8a77a199/log")

  ;; (create cfg :builds
  ;;         {:kind "Build"
  ;;          :apiVersion "ci3.io/v1"
  ;;          :metadata {:name "test-00"}})
  ;; (-> @(http-client/get
  ;;       (url cfg (str "api/v1/namespaces/default/pods"))
  ;;       {:insecure? true
  ;;        :headers (merge default-headers {"Content-Type" "application/json"})})
  ;;     :body
  ;;     (json/parse-string))

  ;; #_(query {:apiVersion "zeroci.io/v1"
  ;;           :ns "default"}
  ;;          :builds "test-1")

  ;; #_(query {:apiVersion "zeroci.io/v1"
  ;;           :ns "default"}
  ;;          :builds "test-1")



  (require '[clj-yaml.core :as yaml])

  (spit "/tmp/pods.yaml"
        (yaml/generate-string
         (query {:prefix "api"
                 :ns "pg3"
                 :apiVersion "v1"}
                :pods)))

  (spit "/tmp/pods.yaml"
        (yaml/generate-string
         (query {:prefix "api"
                 :ns "pg3"
                 :apiVersion "v1"}
                :persistentvolumes)))
  )
(defn query [cfg & pth]
  (println  (apply resource-url cfg pth))
  (let [res @(http-client/get
              (apply resource-url cfg pth)
              {:headers (merge default-headers {"Content-Type" "application/json"})
               :insecure? true})]
    (-> res
     :body
     (json/parse-string keyword)
     (resolve-secrets))))


(defn list [cfg] (query cfg))

(defn find [cfg]
  (query cfg (or (:id cfg) (get-in cfg [:metadata :name]))))

(defn create [res]
  (let [u (resource-url res)]
    (println "POST " u)
    (-> @(http-client/post
          u
          {:body (json/generate-string (walk/stringify-keys res))
           :insecure? true
           :headers (merge default-headers {"Content-Type" "application/json"})})
        :body
        (json/parse-string))))

(defn delete [res]
  (-> @(http-client/delete
        (str (resource-url res (get-in res [:metadata :name])))
        {:headers (merge default-headers {"Content-Type" "application/json"})
         :insecure? true})
      :body
      (json/parse-string)))


(defn *merge
  "merges metadata into one bundle"
  ([a b]
   (loop [[[k v :as i] & ks] b
          acc a]
     (if (nil? i)
       acc
       (let [av (get a k)]
         (if (= v av)
           (recur ks acc)
           (recur ks (if (and (map? v) (map? av))
                       (assoc acc k (*merge av v))
                       (assoc acc k v))))))))
  ([a b & more]
   (apply *merge (*merge a b) more)))

(defn patch [nres]
  (let [res (find nres)]
    (if-not (or (nil? res) (number? res) (= "Failure" (get res :status)))
      (let [diff (patch/diff (walk/stringify-keys res)
                             (walk/stringify-keys (*merge res nres)))]
        (println "RES" res)
        (->
         @(http-client/patch
           (str (resource-url res (get-in res [:metadata :name])))
           {:body (json/generate-string diff)
            :insecure? true
            :headers (merge default-headers {"Content-Type" "application/json-patch+json"})})
         :body))
      (create nres))))

(defn ws-listener
  []
  (reify WebSocketListener
    (onWebSocketConnect [this session]
      (println "Session" session))
    (onWebSocketText [this message]
      (throw (UnsupportedOperationException. "Text not supported")))
    (onWebSocketBinary [this bytes offset len]
      (println "Session" (ByteBuffer/wrap bytes offset len)))
    (onWebSocketError [this cause]
      (when-not (println "not error ch!!!"))
      (when cause
        (println "ERROR:" cause)))
    (onWebSocketClose [this status-code reason]
      (println "on close"))))

(comment
  (query {:kind "Pod"
          :apiVersion "v1"})

  @(http-client/get
    (str kube-url "/api/v1/pods")
    {:headers default-headers :insecure? true})

  (def ws
    (WebSocketClient.
     "wss://<api-server>/api/v1/namespaces/default/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz/exec?command=ls"
     ))


  @(http-client/post
    (str kube-url "/api/v1/namespaces/default/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz/exec?command=ls")
    {:headers default-headers :insecure? true})

  (def ws (WebSocketClient.  (SslContextFactory/trustAnybody)))

  (def req (ClientUpgradeRequest.))

  (def insecure ())

  (.start ws)

  (def conn
    (.connect ws (ws-listener)
              (java.net.URI. "wss://<api-server>/api/v1/namespaces/default/pods/pg3-cleo-master-lightseagreen-363763885-0mbfz/exec?command=ls")
              req))



  )
